src/testing/flow_solver.cpp

```diff
@@ -323,6 +326,41 @@ void FlowSolver<dim,nstate>::output_restart_files(
 }
 #endif
 
+template <int dim, int nstate>
+double FlowSolver<dim,nstate>::calculate_L2_error_at_final_time_wrt_function()
```

Comment:
I think it makes more sense for this to be inside the `time_refinement_study` TestBase class and simply pass a `dg` object and `final_time`, which you can both get from the `flow_solver` object in that class, after doing `flow_solver->run_test()`.

Reply:
I was debating whether this should be in `time_refinement_study` or `flow_solver`, and my logic in putting it here was that it would possibly useful for other tests, therefore putting it in `flow_solver` would make it easier to re-use.

If I were to move this to `time_refinement_study`, I would no longer need to store `exact_solution_function` in the flow solver. This might make it a little easier to ensure that consistent times are compared.

I would like to continue the discussion on this - not sure what is the better way to structure things to make `flow_solver` general but not overcomplicated!

Reply:
Yeah I think it's better to move both `calculate_L2_error_at_final_time_wrt_function()` and `exact_solution_function` to `time_refinement_study`

Reply:
After more thought, I agree. I've moved it in the most recent commit. Thanks for the suggestion!